### PR TITLE
ExtrapolationLineSeries

### LINKED ISSUE #1740: Line Series with style for extrapolated/intropolated regions.
I would suggest a new derived line series to indicate that parts of a curve have been extrapolated/intrapolated. This would be similar to the ThreeColorLine series but the visualization would depend on the horizontal coordinate and not the vertical one. 

I have added an [implementation](https://github.com/Doeharrrck/oxyplot/blob/develop/Source/OxyPlot/Series/ExtrapolationLineSeries.cs) and [examples](https://github.com/Doeharrrck/oxyplot/blob/develop/Source/Examples/ExampleLibrary/Series/ExtrapolationLineSeriesExamples.cs) to my fork of the project.

The examples look like this:
LineFit
Interpolation

Compared to a normal LineSeries, the class ExtrapolationLineSeries has three additional properties:
- Intervals: Gets or sets an IEnumerable<Tuple<double, double>> in which the curve is displayed in extra/interpolated style. Can be set to null. If not null, the intervals must be ascending and may not overlap. Otherwise an InvalidOperationException is thrown. Positive and negative infinity are also legal delimiters. 

- ExtrapolationColor: Gets or sets the color of the intervals in which the curve is visualized as extra/interpolated. In addition the line style is assigned only to these intervals while the rest of the line is in solid style.

- IgnoreExtraplotationForScaling: If set to true, the extrapolated intervals are not taken into account when the minima and maxima of the dataset are calculated. Hence, the extrapolated line will also not scale the axes.

Is it possible to contribute my work?

### PR DESCRIPTION
**Pull request by Doeharrrck**
Fixes #1740 .

### Checklist

- [ x] I have included examples or tests
- [ x] I have updated the change log
- [ x] I am listed in the CONTRIBUTORS file
- [ ?] I have cleaned up the commit history (use rebase and squash)

### Changes proposed in this pull request:

- I have added a series described in issue #1740 and also added several examples.
- The new classes are ExtrapolationLineSeries.cs, ExtrapolationLineSeriesExamples.cs and the support class DataRange.cs plus unit tests.
- I have not modified any existing code except for fixing a comment in RectangleItem.cs.
- I am not sure whether I correctly cleaned up the history because i am new to git.

@oxyplot/admins

### CODE REVIEW

File: CHANGELOG.md
Comment (VisualMelon) [inline CHANGELOG.md:9]: Should mention motivating issue (helps future maintainers find design discussion):

    - ExtrapolationLineSeries and examples (#1740)

File: Source/OxyPlot/Series/DataRange.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/DataRange.cs:108]: Would prefer something like$"[{min}, {max}]", so that it's clearly a closed interval (and distinguishes it from e.g. `ScreenPoint.ToString()`.

File: Source/OxyPlot/Series/DataRange.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/DataRange.cs:30]: Should use proper sentence case in param comments, e.g. "The inclusive lower bound."

File: Source/OxyPlot/Series/DataRange.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/DataRange.cs:25]: - Should be `readonly`.
 - I'd prefer these were non-nullable, and there was an explicit `isDefined` field, and suitable validation was performed in the constructor, but it's not a big deal

File: Source/OxyPlot/Series/DataRange.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/DataRange.cs:34]: I'd prefer that we explicitly reject NaN values here (so that it is impossible to accidently construct an undefined interval), but I'd defer to @objorke for that decision. The rest of the implementation looks fine wrt. to the current behaviour of accepting NaNs.

File: Source/OxyPlot/Series/DataRange.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/DataRange.cs:6]: //   Represents an interval defined by two doubles.

File: Source/OxyPlot/Series/ExtrapolationLineSeries.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/ExtrapolationLineSeries.cs:95]: This means it will use the 'normal' dash style if it was set: we should probably provide another `ExtrapolationLineDashes`.

File: Source/OxyPlot/Series/ExtrapolationLineSeries.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/ExtrapolationLineSeries.cs:100]: Needs tidying a litte.

File: Source/OxyPlot/Series/ExtrapolationLineSeries.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/ExtrapolationLineSeries.cs:159]: Probably not a big deal, but it would be nice to use a binary search for this, so that it scales better with large numbers of intervals.

File: Source/OxyPlot/Series/ExtrapolationLineSeries.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/ExtrapolationLineSeries.cs:310]: max values should be `Math.Max(previous.Maximum, current.Maximum)`, otherwise a wider interval may 'consume' a smaller interval.

**Review by VisualMelon — COMMENTED** [review]
Regarding all the git stuff, basically the idea is that each PR should - in the end - be one trivially mergeable commit. You find more info here: https://github.com/oxyplot/oxyplot/blob/develop/.github/CONTRIBUTING.md#commit-your-changes

I think we can do the squash as part of the merge process, and you your branch doesn't currently need rebasing, though it may become necessary if another PR gets merged in before yours.

File: CHANGELOG.md
Comment (Doeharrrck) [reply CHANGELOG.md]: Done.

File: Source/OxyPlot/Series/DataRange.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/DataRange.cs]: Done.

File: Source/OxyPlot/Series/DataRange.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/DataRange.cs]: Done.

File: Source/OxyPlot/Series/DataRange.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/DataRange.cs]: Done.

File: Source/OxyPlot/Series/DataRange.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/DataRange.cs:40]: Done.

File: Source/OxyPlot/Series/DataRange.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/DataRange.cs]: Done.

File: Source/OxyPlot/Series/ExtrapolationLineSeries.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/ExtrapolationLineSeries.cs]: Done. Also added a new example to test it to the browser.

File: Source/OxyPlot/Series/ExtrapolationLineSeries.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/ExtrapolationLineSeries.cs]: Done.

File: Source/OxyPlot/Series/ExtrapolationLineSeries.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/ExtrapolationLineSeries.cs]: Done. I have added an example with 100,000 points and 1000 intervals to the browser.

File: Source/OxyPlot/Series/ExtrapolationLineSeries.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/ExtrapolationLineSeries.cs]: Fixed.

File: Source/OxyPlot/Series/DataRange.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/DataRange.cs:64]: Maybe not the best name... but I can't think of a better one.

File: Source/OxyPlot/Series/DataRange.cs
Comment (VisualMelon) [inline Source/OxyPlot/Series/DataRange.cs:69]: No longer true

**Comment by VisualMelon** [comment]
I'll take another proper look over the weekend: thanks for your continued work!

File: Source/OxyPlot/Series/DataRange.cs
Comment (Doeharrrck) [reply Source/OxyPlot/Series/DataRange.cs:64]: Measurement studio has a similar class named `Range` where the corresponding property is named 'Interval'. Another alternative might be 'Span'. I personally don't have a strong opinion about which name is best.

File: Source/Examples/ExampleLibrary/Series/ExtrapolationLineSeriesExamples.cs
Comment (VisualMelon) [inline Source/Examples/ExampleLibrary/Series/ExtrapolationLineSeriesExamples.cs:163]: Omits the `[-20, 20]` range.

File: Source/OxyPlot.Tests/Series/DataRangeTests.cs
Comment (VisualMelon) [inline Source/OxyPlot.Tests/Series/DataRangeTests.cs:13]: Unnecessary `usings`

File: Source/OxyPlot.Tests/Series/DataRangeTests.cs
Comment (VisualMelon) [inline Source/OxyPlot.Tests/Series/DataRangeTests.cs:10]: I don't think the `.Series` namespace is necessary: other series specific tests are just thrown in `OxyPlot.Tests`, and the `DataRange` is pretty generic anyway.

**Review by VisualMelon — APPROVED** [review]
Sorry this is a week late.

There are a couple of minor things in the tests and examples, but I think this is basically good to go. I'll wait for @oxyplot or @Jonarw to approve before merging because this adds quite a lot.

**Comment by Doeharrrck** [comment]
Thank you, that's great! I think I have just addressed your last remarks.

**Review by Jonarw — COMMENTED** [review]
Thanks for this PR and sorry for taking so long!
Looks very nice to me, I just have two small nitpicks - other than that, I think this is good to merge.

**Pull request merged**
